mars/oscar/backends/ray/communication.py

```diff
@@ -0,0 +1,281 @@
+        done, _ = await asyncio.wait([self._out_queue.get(), self._closed.wait()],
+                                     return_when=asyncio.FIRST_COMPLETED)
+        if self._closed.is_set():  # pragma: no cover
+            raise ChannelClosed('Channel already closed')
+        if done:
+            return await done.pop()
+
+
+@register_server
+class RayServer(Server):
+    __slots__ = '_closed', '_channels', '_tasks'
+
+    scheme = 'ray'
+
+    def __init__(self, address, channel_handler: Callable[[Channel], Coroutine] = None):
+        super().__init__(address, channel_handler)
+        self._closed = asyncio.Event()
+        self._channels: Dict[ChannelID, RayServerChannel] = dict()
+        self._tasks: Dict[ChannelID, asyncio.Task] = dict()
+
+    @classproperty
+    @implements(Server.client_type)
+    def client_type(self) -> Type["Client"]:
+        return RayClient
+
+    @property
+    @implements(Server.channel_type)
+    def channel_type(self) -> ChannelType:
+        return ChannelType.ray
+
+    @staticmethod
+    @implements(Server.create)
+    async def create(config: Dict) -> "RayServer":
+        config = config.copy()
+        address = config.pop('address')
+        handle_channel = config.pop('handle_channel')
+        if urlparse(address).scheme != RayServer.scheme:  # pragma: no cover
+            raise ValueError(f'Address for RayServer '
+                             f'should be starts with "ray://", '
+                             f'got {address}')
+        if config:  # pragma: no cover
+            raise TypeError(f'Creating RayServer got unexpected '
+                            f'arguments: {",".join(config)}')
+        server = RayServer(address, handle_channel)
+        return server
+
+    @implements(Server.start)
+    async def start(self):
+        # nothing needs to do for ray server
+        pass
+
+    @implements(Server.join)
+    async def join(self, timeout=None):
+        wait_coro = self._closed.wait()
+        try:
+            await asyncio.wait_for(wait_coro, timeout=timeout)
+        except (futures.TimeoutError, asyncio.TimeoutError):
+            pass
+
+    @implements(Server.on_connected)
+    async def on_connected(self, *args, **kwargs):
+        channel = args[0]
+        assert isinstance(channel, RayServerChannel)
+        if kwargs:  # pragma: no cover
+            raise TypeError(f'{type(self).__name__} got unexpected '
+                            f'arguments: {",".join(kwargs)}')
+        await self.channel_handler(channel)
+
+    @implements(Server.stop)
+    async def stop(self):
+        self._closed.set()
+        self._channels = dict()
+        for task in self._tasks.values():
+            task.cancel()
+        self._tasks = dict()
+
+    @property
+    @implements(Server.stopped)
+    def stopped(self) -> bool:
+        return self._closed.is_set()
+
+    async def __on_ray_recv__(self, channel_id: ChannelID, message):
+        channel = self._channels.get(channel_id)
```

Comment:
Is the `__on_ray_recv__` should be blocked if the `self._closed.is_set()`?

Reply:
if the `self._closed.is_set()`, `__on_ray_recv__ ` will throw `ChannelClosed` exception

Reply:
If `__on_ray_recv__` is called after the `RayServer` is stopped, the `self._channels` will be empty, it will create a new session.

Reply:
I think it's better to raise an exception or just pass the `__on_ray_recv__ ` logic here if the `self._closed.is_set()`.